python/cuml/fil/fil.pyx

```diff
@@ -0,0 +1,225 @@
+                                forest_t*,
+                                ModelHandle,
+                                treelite_params_t*)
+
+cdef class FIL_impl():
+
+    cpdef object handle
+    cdef forest_t* forest_pointer
+    cdef forest_t forest_data
+    cdef ModelHandle* model_ptr
+    cdef object depth
+    cdef object num_trees
+    cdef object algo
+    cdef object output
+    cdef object threshold
+    cdef object n_cols
+    cdef object dtype
+    cdef object verbose
+
+    def __cinit__(self, depth=8, n_estimators=50,
+                  output=0, algo=0,
+                  threshold=0.0,
+                  handle=None, verbose=False):
+
+        self.depth = depth
+        self.num_trees = n_estimators
+        self.output = output
+        self.algo = algo
+        self.threshold = threshold
+        self.handle = handle
+        self.verbose = verbose
+        self.forest_pointer = NULL
+
+    def predict(self, X):
+        cdef uintptr_t X_ptr
+        X_m, X_ptr, n_rows, _, X_dtype = \
+            input_to_dev_array(X, order='C')
+
+        cdef cumlHandle* handle_ =\
+            <cumlHandle*><size_t>self.handle.getHandle()
+
+        preds = np.ones(n_rows, dtype=np.float32)
+        cdef uintptr_t preds_ptr
+        preds_m, preds_ptr, _, _, _ = \
+            input_to_dev_array(preds)
+        print( " dtype of X : ", X_dtype)
+        #cdef forest_t forest_data = <forest_t> &self.forest_pointer
+        predict(handle_[0],
+                self.forest_data,
+                <float*> preds_ptr,
+                <float*> X_ptr,
+                <size_t> n_rows)
+        self.handle.sync()
+        # synchronous w/o a stream
+        preds = preds_m.copy_to_host()
+        return preds
+
+    def free(self):
+
+        cdef cumlHandle* handle_ =\
+            <cumlHandle*><size_t>self.handle.getHandle()
+
+        free(handle_[0],
+             self.forest_data)
+
+        return self
+
+    def from_treelite(self, model, output_class):
+
+        cdef treelite_params_t treelite_params
+        treelite_params.output_class = output_class
+        treelite_params.threshold = self.threshold
+        treelite_params.algo = self.algo
+        #model_copy = copy.deepcopy(model)
+        #cdef ModelHandle model_ptr = new ModelHandle*> &model
+        print(" model.num_feature : ", model.num_feature)
+        print(" model.num_output_group : ",model.num_output_group)
+        print(" model handle : ", model.handle)
+        self.forest_pointer = <forest_t*><size_t> model.handle.value
+        cdef cumlHandle* handle_ =\
+            <cumlHandle*><size_t>self.handle.getHandle()
+        cdef uintptr_t model_ptr = model.handle.value
+        self.forest_data = from_treelite(handle_[0],
+                                         self.forest_pointer,
+                                         <ModelHandle> model_ptr,
+                                         &treelite_params)
+
+        return self
+
+class FIL(Base):
```

Comment:
Are two classes really needed here? A single one should be enough.

Reply:
we need two classes since FIL is stateful